src/Icicle/Source/Checker/Constraint.hs

```diff
@@ -93,21 +103,25 @@ constraintsQ env q
 -- The constraints flow upwards, bottom-up, but are also discharged and simplified at every level.
 -- The substitutions from discharged constraints also flow upwards.
 --
--- The environment mapping flows downwards, with new names only available in lexically nested subexpressions and so on.
+-- The environment mapping flows downwards, with new names only available in lexically
+-- nested subexpressions and so on.
 --
--- Shoehorning these two differently flowing contexts into a state monad was a mistake, and is a constant source of bugs.
---
-generateQ :: Ord n => Query a n -> Gen a n (Query'C a n, SubstT n)
+generateQ
+  :: Ord n
+  => Query a n
+  -> CheckContext a n
+  -> Gen a n (Query'C a n, SubstT n, CheckContext a n)
```

Comment:
yes, I wasn't quite sure if I would not need the Environment anywhere, so I make it keep both the Environment + Constraint like the state monad was doing. Now I had a better idea so I'll only return the constraint set. Sorry, so clueless =(